hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDBCheckpointDiffer.java

```diff
@@ -939,6 +980,290 @@ private void populateCompactionDAG(List<String> inputFiles,
+
+    while (!currentLevel.isEmpty()) {
+      Set<CompactionNode> nextLevel = new HashSet<>();
+      for (CompactionNode current : currentLevel) {
+        if (!forwardDag.nodes().contains(current)) {
+          continue;
+        }
+
+        nextLevel.addAll(forwardDag.successors(current));
+        forwardDag.removeNode(current);
+        removedFiles.add(current.getFileName());
+      }
+
+      currentLevel = nextLevel;
+    }
+
+    return removedFiles;
+  }
+
+  private SnapshotLogInfo getSnapshotInfoFromLog(Path compactionLogFile) {
+    AtomicReference<SnapshotLogInfo> snapshotLogInfo =
+        new AtomicReference<>();
+    try (Stream<String> logStream = Files.lines(compactionLogFile, UTF_8)) {
+      logStream.forEach(logLine -> {
+        if (!logLine.startsWith(COMPACTION_LOG_SEQ_NUM_LINE_PREFIX)) {
+          return;
+        }
+
+        snapshotLogInfo.set(getSnapshotLogInfo(logLine));
+      });
+    } catch (IOException exception) {
+      throw new RuntimeException("Failed to read compaction log file: " +
+          compactionLogFile, exception);
+    }
+
+    return snapshotLogInfo.get();
+  }
+
+  /**
+   * Converts a snapshot compaction log line to SnapshotLogInfo.
+   */
+  private SnapshotLogInfo getSnapshotLogInfo(String line) {
+    String[] splits = line.split(" ");
+    assert (splits.length == 4);
+
+    return new SnapshotLogInfo(Long.parseLong(splits[1]),
+        splits[2],
+        Long.parseLong(splits[3]));
+  }
+
+  /**
+   * Returns the list of SST files got compacted in the last compaction from
+   * the provided list of compaction log files.
+   * We can't simply use last file from the list because it is possible that
+   * no compaction happened between the last snapshot and previous to that.
+   * Hence, we go over the list in reverse order and return the SST files from
+   * first the compaction happened in the reverse list.
+   * If no compaction happen at all, it returns empty list.
+   */
+  @VisibleForTesting
+  List<String> getLastCompactionSstFiles(
+      List<Path> compactionLogFiles
+  ) {
+
+    if (compactionLogFiles.isEmpty()) {
+      return Collections.emptyList();
+    }
+    compactionLogFiles = new ArrayList<>(compactionLogFiles);
+    Collections.reverse(compactionLogFiles);
+
+    for (Path compactionLogFile: compactionLogFiles) {
+      List<String> sstFiles = getLastCompactionSstFiles(compactionLogFile);
+      if (!sstFiles.isEmpty()) {
+        return  sstFiles;
+      }
+    }
+
+    return Collections.emptyList();
+  }
+
+  private List<String>  getLastCompactionSstFiles(Path compactionLogFile) {
+
+    AtomicReference<String> sstFiles = new AtomicReference<>();
+
+    try (Stream<String> logStream = Files.lines(compactionLogFile, UTF_8)) {
+      logStream.forEach(logLine -> {
+        if (!logLine.startsWith(COMPACTION_LOG_ENTRY_LINE_PREFIX)) {
+          return;
+        }
+        sstFiles.set(logLine);
+      });
```

Comment:
This is a creative way to locate the last "C " compaction line. I wonder how costly atomic set is tho.